<a href="https://colab.research.google.com/github/MarisovaTanya/Final-work_python/blob/main/Final_work.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd

In [2]:
df0 = pd.read_csv('HR.csv')
df0.head()

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2                   157   
1                0.80             0.86               5                   262   
2                0.11             0.88               7                   272   
3                0.72             0.87               5                   223   
4                0.37             0.52               2                   159   

   time_spend_company  Work_accident  left  promotion_last_5years department  \
0                   3              0     1                      0      sales   
1                   6              0     1                      0      sales   
2                   4              0     1                      0      sales   
3                   5              0     1                      0      sales   
4                   3              0     1                      0      sales   

   salary  
0     low  
1  medium  
2  medium  
3     low  
4     low

In [3]:
df = df0.drop_duplicates().reset_index(drop=True)
df.head()

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2                   157   
1                0.80             0.86               5                   262   
2                0.11             0.88               7                   272   
3                0.72             0.87               5                   223   
4                0.37             0.52               2                   159   

   time_spend_company  Work_accident  left  promotion_last_5years department  \
0                   3              0     1                      0      sales   
1                   6              0     1                      0      sales   
2                   4              0     1                      0      sales   
3                   5              0     1                      0      sales   
4                   3              0     1                      0      sales   

   salary  
0     low  
1  medium  
2  medium  
3     low  
4     low

2. Рассчитайте основные статистики для переменных
(среднее,медиана,мода,мин/макс,сред.отклонение).

In [4]:
df.iloc[:,:5].describe().round(2).loc[['50%', 'mean', 'min', 'max', 'std']]

satisfaction_level  last_evaluation  number_project  \
50%                 0.66             0.72            4.00   
mean                0.63             0.72            3.80   
min                 0.09             0.36            2.00   
max                 1.00             1.00            7.00   
std                 0.24             0.17            1.16   

      average_montly_hours  time_spend_company  
50%                 200.00                3.00  
mean                200.47                3.36  
min                  96.00                2.00  
max                 310.00               10.00  
std                  48.73                1.33

In [5]:
df.iloc[:,5:10].mode()

Work_accident  left  promotion_last_5years department salary
0              0     0                      0      sales    low

3. Рассчитайте и визуализировать корреляционную матрицу для
количественных переменных.
Определите две самые скоррелированные и две наименее
скоррелированные переменные

In [6]:
df_corrs = df[['satisfaction_level', 'last_evaluation', 'number_project', 'average_montly_hours', 'time_spend_company']].corr().round(3)
df_corrs

satisfaction_level  last_evaluation  number_project  \
satisfaction_level                 1.000            0.095          -0.133   
last_evaluation                    0.095            1.000           0.270   
number_project                    -0.133            0.270           1.000   
average_montly_hours              -0.006            0.265           0.332   
time_spend_company                -0.153            0.097           0.189   

                      average_montly_hours  time_spend_company  
satisfaction_level                  -0.006              -0.153  
last_evaluation                      0.265               0.097  
number_project                       0.332               0.189  
average_montly_hours                 1.000               0.103  
time_spend_company                   0.103               1.000

In [7]:
df_corrs.style.background_gradient(cmap='RdYlGn')

Наиболее скореллированные переменные: average_montly_hours и number_project - 0.332

Наименее скоррелированные переменные average_montly_hours и satisfaction_level - (-0.006)

4. Рассчитайте сколько сотрудников работает в каждом
департаменте.

In [8]:
df['department'].value_counts()

department
sales          3239
technical      2244
support        1821
IT              976
RandD           694
product_mng     686
marketing       673
accounting      621
hr              601
management      436
Name: count, dtype: int64

5. Показать распределение сотрудников по зарплатам.

In [9]:
df['salary'].value_counts()

salary
low       5740
medium    5261
high       990
Name: count, dtype: int64

6. Показать распределение сотрудников по зарплатам в каждом
департаменте по отдельности

In [10]:
df.pivot_table(index='department', columns='salary', aggfunc='size')

salary       high   low  medium
department                     
IT             71   476     429
RandD          47   322     325
accounting     63   296     262
hr             38   296     267
management    128   139     169
marketing      62   310     301
product_mng    52   343     291
sales         237  1553    1449
support       126   867     828
technical     166  1138     940

7. Проверить гипотезу, что сотрудники с высоким окладом
проводят на работе больше времени, чем сотрудники с низким
окладом

Гипотеза H0 - сотрудники с высоким окладом
проводят на работе столько же времени, что и сотрудники с низким
окладом

In [11]:
salary_low = df[df['salary'] == 'low']['average_montly_hours']
salary_high = df[df['salary'] == 'high']['average_montly_hours']

In [12]:
import scipy.stats as stats

In [13]:
result = stats.ttest_ind(salary_low, salary_high, equal_var=False)

a = 0.05

if (result.pvalue<a):
    print('Отвергаем нулевую гипотезу')
else:
    print('Не отвергаем нулевую гипотезу')

Не отвергаем нулевую гипотезу


Результат:сотрудники с высоким окладом проводят на работе столько же времени, что и сотрудники с низким окладом

8. Рассчитать следующие показатели среди уволившихся и не
уволившихся сотрудников (по отдельности):

● Доля сотрудников с повышением за последние 5 лет

● Средняя степень удовлетворенности

● Среднее количество проектов

In [14]:
working= df[(df['left'] == 0)]
fired =  df[(df['left'] == 1)]

In [16]:
table = []

table.append(
    {
    'Показатель': 'Доля сотрудников с повышением за последние 5 лет',
    'Уволившиеся': round(fired.loc[fired.promotion_last_5years==1,'promotion_last_5years'].count()/fired['promotion_last_5years'].count(),3),
    'Действующие': round(working.loc[working.promotion_last_5years==1,'promotion_last_5years'].count()/working['promotion_last_5years'].count(),3)
    }
)

table.append(
    {
    'Показатель': 'Средняя степень удовлетворенности',
    'Уволившиеся': round(fired.satisfaction_level.mean(), 3),
    'Действующие': round(working.satisfaction_level.mean(), 3)
    }
)

table.append(
    {
    'Показатель': 'Среднее количество проектов',
    'Уволившиеся': round(fired.number_project.mean(), 3),
    'Действующие': round(working.number_project.mean(), 3)
    }
)
table

[{'Показатель': 'Доля сотрудников с повышением за последние 5 лет',
  'Уволившиеся': 0.004,
  'Действующие': 0.02},
 {'Показатель': 'Средняя степень удовлетворенности',
  'Уволившиеся': 0.44,
  'Действующие': 0.667},
 {'Показатель': 'Среднее количество проектов',
  'Уволившиеся': 3.883,
  'Действующие': 3.787}]

9. Разделить данные на тестовую и обучающую выборки.
Построить модель LDA, предсказывающую уволился ли
сотрудник на основе имеющихся факторов (кроме department и
salary).
Оценить качество модели на тестовой выборки

In [36]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split

In [28]:
X = df.iloc[:,[0,1,2,3,4,5,7]]
y = df['left']

In [29]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=42)

In [30]:
model = LinearDiscriminantAnalysis()
model.fit(X_train, y_train)

LinearDiscriminantAnalysis()

In [31]:
model.coef_

array([[-4.41846686e+00,  7.85449186e-01, -1.83813535e-01,
         4.40193080e-03,  3.00782062e-01, -9.31516558e-01,
        -9.79941863e-01]])

In [32]:
model.intercept_

array([-0.88848447])

In [33]:
y_pred = model.predict(X_test)
y_pred

array([1, 0, 0, ..., 0, 1, 0])

In [34]:
model.score(X_test, y_test) # Коэффициент детерминации - 0.83, а это означает, что модель получилась хорошая

0.830739299610895

In [37]:
accuracy_score(y_test, model.predict(X_test)) # Точность модели высокая

0.830739299610895